### PR TITLE
Tweaks and small fixes to the prototype

### PR DESCRIPTION
**Pull request by mheinzel**
Probably easiest to review commit by commit.

I managed to find a few invariants that too tight, or looser than necessary. However, this required adding handwritten tests, as the lockstep test coverage is not great.

The fix for incorrect assumptions violated by 5-way merges (holding back an underfull run) are simply addressed by providing more merge credits, but there are alternatives we can think about afterwards (e.g. #311)

### CODE REVIEW

File: prototypes/ScheduledMerges.hs [hunk 1]
```diff
@@ -163,101 +159,115 @@ mergeLastForLevel _  = MergeMidLevel
...
         -- (almost) always involved in an ongoing merge. Thus there are no
         -- other "normal" runs. The exception is when a levelling run becomes
         -- too large and is promoted, in that case initially there's no merge,
         -- but it is still represented as a 'MergingRun', using 'SingleRun'.
-        MergePolicyLevelling -> null rs
-        MergePolicyTiering   -> all (\r -> tieringRunSizeToLevel r == ln) rs
+        MergePolicyLevelling -> assertST $ null rs
+        -- Runs in tiering levels usually fit that size, but they can be one
+        -- larger, if a run has been held back (creating a 5-way merge).
+        MergePolicyTiering   -> assertST $ all (\r -> tieringRunSizeToLevel r `elem` [ln, ln+1]) rs
```
Comment (mheinzel) [inline prototypes/ScheduledMerges.hs:197]: This is actually still not really true (although will hold in practice).

In the pathological case, due to holding back, all runs passed to the next level can be factor (5/4) too large, and there the same holding back can happen, leading to factor (6/4) etc., until 12 levels down a run is *two* levels too large.

File: prototypes/ScheduledMerges.hs [hunk 2]
```diff
@@ -281,13 +277,15 @@ newMerge tr level mergepolicy mergelast rs = do
   where
     cost = sum (map Map.size rs)
     -- How much we need to discharge before the merge can be guaranteed
-    -- complete.
+    -- complete. More precisely, this is the maximum amount a merge at this
+    -- level could need. This overestimation means that merges will only
+    -- complete at the last possible moment.
```
Comment (jorisdral) [inline prototypes/ScheduledMerges.hs:297]: Well, that depends: the merge can finish earlier, but it should never finish too late. Maybe reword this a little bit?

File: prototypes/ScheduledMergesTestQLS.hs [hunk 3]
```diff
@@ -84,6 +81,48 @@ test_regression_empty_run =
...
+-- That merge gets completed by supplying credits.
+test_merge_again_with_incoming :: IO ()
+test_merge_again_with_incoming =
+    runWithTracer $ \tracer -> do
+      stToIO $ do
+        lsm <- LSM.new
+        let ins k = LSM.insert tracer lsm k 0
+        -- get something to 3rd level (so 2nd level is not levelling)
+        -- (needs 5 runs to go to level 2 so the resulting run becomes too big)
+        traverse_ ins [101..100+(5*16)]
```
Comment (jorisdral) [inline prototypes/ScheduledMergesTestQLS.hs:95]: Let's record assumptions like this about the structure in an assertion. i.e., check that there is a run on the third level. This should ensure that the test can't become outdated if we change the prototype. And let's do that in other places too where it makes sense, not just this line

Comment (jorisdral) [reply prototypes/ScheduledMergesTestQLS.hs:95]: Optionally, you could also make this test size factor independent

File: prototypes/ScheduledMergesTestQLS.hs [hunk 4]
```diff
@@ -84,6 +81,48 @@ test_regression_empty_run =
...
+        -- get a very small run (4 elements) to 2nd level
+        replicateM_ 4 $
+          traverse_ ins [201..200+4]
```
Comment (jorisdral) [inline prototypes/ScheduledMergesTestQLS.hs:98]: Likewise, assert that the resulting run is too small

File: prototypes/ScheduledMergesTestQLS.hs [hunk 5]
```diff
@@ -84,6 +81,48 @@ test_regression_empty_run =
...
+        -- get another run to 2nd level, which the small run can be merged with
+        traverse_ ins [301..300+16]
+        -- complete the merge
+        LSM.supply lsm 32
```
Comment (jorisdral) [inline prototypes/ScheduledMergesTestQLS.hs:102]: Should it not complete in 16 steps?

File: prototypes/ScheduledMergesTestQLS.hs [hunk 6]
```diff
@@ -84,6 +81,48 @@ test_regression_empty_run =
...
+        -- (needs 5 runs to go to level 2 so the resulting run becomes too big)
+        traverse_ ins [101..100+(5*16)]
+        -- get a very small run (4 elements) to 2nd level
+        replicateM_ 4 $
+          traverse_ ins [201..200+4]
+        -- get another run to 2nd level, which the small run can be merged with
+        traverse_ ins [301..300+16]
+        -- get 3 more to 2nd level, so the merge above is expected to complete
+        -- (actually more, as runs only move once a fifth run arrives...)
+        traverse_ ins [401..400+(6*16)]
```
Comment (jorisdral) [inline prototypes/ScheduledMergesTestQLS.hs:123]: Likewise, aren't we doing 16 steps too many?

File: prototypes/ScheduledMergesTestQLS.hs [hunk 7]
```diff
@@ -209,6 +209,13 @@ instance StateModel (Lockstep Model) where
     ADuplicate :: ModelVar Model (LSM RealWorld)
                -> Action (Lockstep Model) (LSM RealWorld)
 
+    -- | Without this, the prototype only completes tiering merges when the next
+    -- merging run on this level is created, so a level would never contain a
+    -- completed merge.
+    ASupply :: ModelVar Model (LSM RealWorld)
+            -> Int
+            -> Action (Lockstep Model) ()
```
Comment (jorisdral) [inline prototypes/ScheduledMergesTestQLS.hs:217]: It's not clear to me why this is necessary. Shouldn't the prototype be able to maintain it's invariants without supplying credits explicitly? If we add this action, we're not letting the prototype "do its thing"

File: prototypes/ScheduledMerges.hs [hunk 8]
```diff
@@ -466,20 +466,24 @@ bufferToRun = id
...
-creditsForMerge (MergingRun MergePolicyLevelling _ _) = 5
-creditsForMerge (MergingRun MergePolicyTiering   _ _) = 1
+-- A levelling merge has 1 input run and one resident run, which is (up to) 4x
+-- bigger than the others.
+-- It needs to be completed before another run comes in.
+creditsForMerge (MergingRun MergePolicyLevelling _ _) = (1 + 4) / 1
+
+-- A tiering merge has 5 runs at most (once could be held back to merged again)
+-- and must be completed before the level is full (once 4 more runs come in).
+creditsForMerge (MergingRun MergePolicyTiering   _ _) = 5 / 4
```
Comment (jorisdral) [inline prototypes/ScheduledMerges.hs:486]: This could take into account whether it is merging 4 runs or 5. In the former case we'd supply 1 credit, and 1.25 in the latter case

**Review by jorisdral — COMMENTED** [review]
Looks promising, I've left some comments but haven't looked into all changes

File: prototypes/ScheduledMerges.hs (hunk 2 above)
Comment (mheinzel) [reply prototypes/ScheduledMerges.hs:297]: In a real implementation that's true. But in the prototype merges will never finish early, as their completion is not based on the any "real" work to be performed (which would depend on the number of entries), but on the debt calculated here, which is the maximum a merge at this level could require.

I can add that explanation to the comment, though.

File: prototypes/ScheduledMergesTestQLS.hs (hunk 5 above)
Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:102]: I think you're right, I just supplied more as the batch size is 32 (and supplying too much doesn't hurt). But 16 should work, independent of the batch size. :+1:

File: prototypes/ScheduledMergesTestQLS.hs (hunk 7 above)
Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:217]: Without this, the whole part of the invariant that checks a `CompletedMerge` present in a tiering level is dead code. Runs stay `Ongoing` until they are replaced by the next incoming merge.

Maybe that's something I could change in the prototype, related to https://github.com/IntersectMBO/lsm-tree/pull/310#discussion_r1696639389.

File: prototypes/ScheduledMerges.hs (hunk 2 above)
Comment (mheinzel) [reply prototypes/ScheduledMerges.hs:297]: I think the reasoning for always requiring the maximum amount of work is that it makes it more likely to find issues related to supplying too few credits somewhere. We could base everything on `cost` (sum of input run sizes), but not sure whether it's an improvement or not.

File: prototypes/ScheduledMerges.hs (hunk 8 above)
Comment (mheinzel) [reply prototypes/ScheduledMerges.hs:486]: Good point, I'll make that change.

File: prototypes/ScheduledMergesTestQLS.hs (hunk 7 above)
Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:217]: Without `ASupply`, I can do

```diff
--- a/prototypes/ScheduledMerges.hs
+++ b/prototypes/ScheduledMerges.hs
@@ -247,6 +247,7 @@ invariant = go 1
             -- Note that tiering on the last level only occurs when there is
             -- a single level only.
             (_, CompletedMerge r, MergeLastLevel) -> do
+              assertST $ False
               assertST $ ln == 1
               assertST $ tieringRunSizeToLevel r <= ln+1
 
@@ -254,7 +255,8 @@ invariant = go 1
             -- level it is entering, but can also be one smaller (in which case
             -- it'll be held back and merged again) or one larger (because it
             -- includes a run that has been held back before).
-            (_, CompletedMerge r, MergeMidLevel) ->
+            (_, CompletedMerge r, MergeMidLevel) -> do
+              assertST $ False
               assertST $ tieringRunSizeToLevel r `elem` [ln-1, ln, ln+1]
```

And even increase the generator size a lot and don't get any failures:

```
prototype
  ScheduledMerges
    ScheduledMerges vs model: OK (88.31s)
      +++ OK, passed 100 tests.
      
      Action polarity (83563 in total):
      100.000% +
      
      Actions (83563 in total):
      33.441% +AInsert
      32.943% +ALookup
      16.808% +ADelete
       8.383% +ADump
       8.308% +ADuplicate
       0.118% +ANew
```

Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:217]: Although arguably this just putting on a bandaid. The lockstep tests just don't reach proper coverage. For the *first* tiering merge to even be created, ~100 elements need to be inserted, which means a lot of actions are required before anything interesting happens.

Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:217]: I'll remove `ASupply` and instead make sure the invariant gets checked inbetween supplying credits and doing the other work (e.g. requiring a merge to be completed), so it has a chance to look at every merge once it completes. The need for better generators remains, though.

Comment (jorisdral) [reply prototypes/ScheduledMergesTestQLS.hs:217]: One could add an action that performs multiple inserts (instead of just 1). Would that hit the corner case?

**Comment by jorisdral** [comment]
Posting this here so I don't forget: I was reading the Monkey paper and found this quote:

> If multiple runs that are being sort-merged contain entries with the same key, only the entry
from the most recently-created (youngest) run is kept because it is the most up-to-date. Thus,
the resulting run may be smaller than the cumulative sizes of the original runs. When a merge
operation is finished, the resulting run moves to Level i + 1 if Level i is at capacity.

Specially the last part we do differently: we look *only* at the size of the run to determine whether it should stay, and we don't look at the overall level capacity. Would it be a reasonable alternative to promote runs based on the level capacity, not just their own size?

**Comment by mheinzel** [comment]
> Would it be a reasonable alternative to promote runs based on the level capacity, not just their own size?

Ah interesting. Not sure I fully understand it (maybe it's just talking about levelling, where this is pretty much what we do already?), but should be possible for tiering as well. However, it weakens the invariants on run sizes quite a bit.

File: prototypes/ScheduledMerges.hs [hunk 9]
```diff
@@ -222,15 +225,20 @@ invariant = go 1
             -- same reasons as above. Although if this is the first merge for
             -- a new level, it'll have only 4 runs.
             (_, OngoingMerge _ rs _) -> do
-              assertST $ length rs == 4 || length rs == 5
-              assertST $ all (\r -> tieringRunSizeToLevel r `elem` [ln-1, ln]) (take 4 rs)
-              assertST $ all (\r -> levellingRunSizeToLevel r <= ln+1) (drop 4 rs)
+              let incoming = take 4 rs
+              let resident = drop 4 rs
+              assertST $ length incoming == 4
+              assertST $ length resident <= 1
```
Comment (dcoutts) [inline prototypes/ScheduledMerges.hs:231]: Ok, ok, this is indeed a funny but correct way of saying that length rs is 4 or 5. "the bit less than or equal to 4 must be exactly 4 and the bit more than 4 must be at most 1". :exploding_head: 

I don't think it's clearer!

**Review by dcoutts — APPROVED** [review]
Yes, this isn't horrible, but does show some things that are a bit horrible (the possibility of ending up with too-big runs as well as too-small).

I also endorse the comments from Joris.

File: prototypes/ScheduledMerges.hs [hunk 10]
```diff
@@ -159,101 +159,103 @@ mergeLastForLevel _  = MergeMidLevel
...
-                (length rs == 4 || length rs == 5)
-             && all (\r -> tieringRunSizeToLevel r == ln-1) rs
+            (_, OngoingMerge _ rs _, _) -> do
+              assertST $ length rs == 4 || length rs == 5
+              assertST $ all (\r -> tieringRunSizeToLevel r == ln-1) rs
+
+-- 'callStack' just ensures that the 'HasCallStack' constraint is not redundant
+-- when compiling with debug assertions disabled.
+assertST :: HasCallStack => Bool -> ST s ()
+assertST p = assert p $ return (const () callStack)
```
[the remark this answers is not in the document]
Comment (mheinzel) [reply prototypes/ScheduledMerges.hs:270]: With `assertST p = assert p $ return ()` I got this: https://github.com/IntersectMBO/lsm-tree/actions/runs/10092810012/job/27907230850#step:13:201

File: prototypes/ScheduledMerges.hs (hunk 9 above)
Comment (mheinzel) [reply prototypes/ScheduledMerges.hs:231]: I think it helps in understanding why those two cases are possible. But since the comment above explains it as well, I can change it to ``length rs `elem` [4, 5]``. :+1: 
The important part is that `incoming` and `resident` have different invariants in the checks in the following two lines.

File: prototypes/ScheduledMergesTestQLS.hs (hunk 3 above)
Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs:95]: I decided not to make it size factor independent for now, but I think I found a decent way of making assertions about the LSM shape. It's a bit annoying that merges always get created a little later than one (at least I) would think, e.g. flushing 4 write buffers does not yet result in a merge, but that's just how the algorithm currently works (and also the reason for losing some potential sharing).

File: prototypes/ScheduledMergesTestQLS.hs (hunk 7 above)
Comment (mheinzel) [reply prototypes/ScheduledMergesTestQLS.hs]: I tried it, but it didn't catch any of the incorrect invariants I found with the handwritten tests. My implementation was not ideal, though, it can probably be improved. I only return one of the inserted keys (so it gets added to the available variables), which means key re-use is less likely and it's harder to get a lot of compaction. Returning multiple keys from a single action turned out to be tricky, so I abandoned that.

**Comment by mheinzel** [comment]
> Maybe, if you can fit in [#310 (comment)](https://github.com/IntersectMBO/lsm-tree/pull/310#issuecomment-2284084875), that would be nice, otherwise could you create an issue for it that links to my comment?

I can open an issue and quickly outline various ideas for changing when to start a merge. Thinking about that again, "if Level i is at capacity" might just mean "if there are 4 runs", it doesn't say whether capacity refers to number of entries or number of runs. But I should have another look at the paper at some point.

**Comment by mheinzel** [comment]
I adressed the feedback and rebased. What do you think?

**Comment by jorisdral** [comment]
> > Maybe, if you can fit in [#310 (comment)](https://github.com/IntersectMBO/lsm-tree/pull/310#issuecomment-2284084875), that would be nice, otherwise could you create an issue for it that links to my comment?
> 
> I can open an issue and quickly outline various ideas for changing when to start a merge. 

Yeah, let's open an issue

> Thinking about that again, "if Level i is at capacity" might just mean "if there are 4 runs", it doesn't say whether capacity refers to number of entries or number of runs. But I should have another look at the paper at some point.

No, capacity talks about the entries in the run, not the number of runs

**Review by jorisdral — APPROVED** [review]
